### PR TITLE
common/config: add ConfigProxy for crimson

### PR DESCRIPTION
**Pull request by tchaikov**

### CODE REVIEW

**Comment by tchaikov** [comment]
i have not finished the `crimson/common/config_proxy.*` and its test. this PR is mainly for getting early review.

**Comment by tchaikov** [comment]
@cbodley @jdurgin ping?

**Comment by tchaikov** [comment]
@liu-chunmei in case you are interested.

File: src/common/config_obs_mgr.h [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+  }
+  assert(found_obs);
+}
+
+template<class ConfigObs>
+template<class ConfigProxyT>
+void ObserverMgr<ConfigObs>::call_all_observers(const ConfigProxyT& proxy)
+{
+  rev_obs_map_t rev_obs;
+  // Have the scope of the lock extend to the scope of
```
Comment (jdurgin) [inline src/common/config_obs_mgr.h]: nit: locking comment is no longer accurate with crimson

Comment (tchaikov) [reply src/common/config_obs_mgr.h]: true, moved it to `src/common/config_proxy.h`, where we have the "lock".

**Comment by tchaikov** [comment]
http://pulpito.ceph.com/kchai-2018-07-25_04:36:50-rados-wip-seastar-config-kefu-distro-basic-mira/

File: src/crimson/common/config_proxy.h [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    return remote_config ? *remote_config : * local_config;
+  }
+
+  // apply changes to all shards
+  // @param func a functor which accepts @c "ConfigValues&", and returns
+  //             a boolean indicating if the values is changed or not
+  template<typename Func>
+  seastar::future<> do_change(Func&& func) {
+    auto new_values = seastar::make_lw_shared(*values);
+    func(*new_values);
```
Comment (cbodley) [inline src/crimson/common/config_proxy.h]: can we check the return value of `func()` for false here to skip the rest of the function if nothing changed?

File: src/crimson/common/config_proxy.h [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    return seastar::do_with(seastar::make_foreign(new_values),
+			    [this](auto& foreign_values) {
+      return container().invoke_on_all([&foreign_values](ConfigProxy& proxy) {
+        return foreign_values.copy().then([&proxy](auto foreign_values) {
+	  proxy.values.reset();
+	  proxy.values = std::move(foreign_values);
```
Comment (cbodley) [inline src/crimson/common/config_proxy.h]: i'm having trouble following the design here. it looks like concurrent calls to `do_change()` will race (even from the same shard) and result in one change overwriting the others. take for example:
```c++
auto fut1 = set_val("A", "1");
auto fut2 = set_val("B", "2");
...
return seastar::when_all(std::move(fut1), std::move(fut2)); // wait for both
```
both calls to `set_val()` will copy the initial values and apply a mutation. each shard will assign the values with A=1 first, then overwrite that with the copy containing B=2. but there's no way to end up with both A=1 and B=2 without additional synchronizaion

maybe you could apply the local values synchronously first, then use `invoke_on_others()` to apply the rest - that would at least prevent calls on the same shard from racing. after that, you could choose shard 0 as the only writer, and proxy all calls to `do_change()` through that shard first?

alternatively..

if each shard had its own local copy of the `ConfigValues`, `do_change()` could pass the `func` itself to each shard and mutate its values in-place. this strategy would probably still require choosing a single shard to call `invoke_on_all()` to preserve the order of mutations, but would save us from having to do gc with lw_shared_ptr/foreign_ptr

**Comment by cbodley** [comment]
sorry for late review, i didn't realize you'd merged!

File: src/crimson/common/config_proxy.h (hunk 3 above)
Comment (tchaikov) [reply src/crimson/common/config_proxy.h]: @cbodley i don't think it's a usecase we want to address here. the consumer of `ceph::common::ConfigProxy` is seastar-osd. for updating configs, osd has following usecases:

- `injectargs()`
- `set_val()` then `apply_changes()`
- `rm_val()` then `apply_changes()`
- if `osd_debug_shutdown`, `set_val()`, `set_val()`, ... then `apply_changes()`. if we really want this in seastar-osd, we could chain them.

in other words, we don't change the settings in batch from a certain shard. 

yes, the concurrent calls to `do_change()` race both between shards and on a single shard. and to serialize the write ops using shard.0 was the first design i have in my mind. but after a second thought, i think it's fine to have this racing. as, in general, the settings are not updated by multiple writers in parallel. and it'd be symmetric and is simpler if we use current design.

but if you insist, i am happy to revise it following the first design suggested by you. it practically makes the shard.0 the leader of the sharded service, put in the ceph's paxos variant's way. i don't really want to let each shard have its own copy of `ConfigValues`, as i assume the `ConfigValues` is a relatively big data structure, which should be shared among shards.

File: src/common/config_proxy.h [hunk 4]
```diff
@@ -55,16 +66,20 @@ class ConfigProxy {
   }
   void diff(Formatter *f, const std::string& name=string{}) const {
     return config.diff(values, f, name);
+    Mutex::Locker l{lock};
```
Comment (liu-chunmei) [inline src/common/config_proxy.h]: here should lock first then call config.diff, right?

File: src/crimson/common/config_proxy.h (hunk 3 above)
Comment (tchaikov) [reply src/crimson/common/config_proxy.h]: will send a follow-up PR very soon.

File: src/common/config_proxy.h (hunk 4 above)
Comment (tchaikov) [reply src/common/config_proxy.h]: ahh, yeah! thanks.

Comment (tchaikov) [reply src/common/config_proxy.h]: #23276

**Pull request merged**